### ISSUE TITLE
M1: Uplift to TS 26.512 V17.7.0

### ISSUE THREAD

**Issue by rjb1000** [5G-MAG/rt-5gms-application-function#112]
# Background

The following changes to TS 26.512 were agreed at SA4#126 (Chicago) and approved at SA#102 (Edinburgh):

* CR0054 "Essential maintenance" in [S4-231631](https://www.3gpp.org/ftp/TSG_SA/WG4_CODEC/TSGS4_126_Chicago/Docs/S4-231631.zip) and plenary CR Pack [SP-231368](https://www.3gpp.org/ftp/TSG_SA/TSG_SA/TSGS_102_Edinburgh_2023-12/Docs/SP-231368.zip).

The updated specification will be published shortly as TS 26.512 V17.7.0.

This issue captures the changes that need to be made to the reference implementation of the 5GMS AF Provisioning API at reference point M1.

# Summary of changes

1.  _applicationSessionContext_ becomes an optional property of the Policy Template resource in **Policy Templates Provisioning API**.
2. **Policy Templates Provisioning API** now has `readOnly: true`  for _policyTemplateId_, _state_ and _stateReason_.
3. HTTP response code changed for malformed regular expression error in cache purge operation of **Content Hosting Provisioning API** (clauses 7.6.4.3 and C.3.5).
    - *422 (Unprocessable Content)* becomes *400 (Bad Request)*.
4. **Consumption Reporting Provisioning API** enforces boundaries for _reportingInterval_.
    -  `minimum: 0`.
    -  `exclusiveMinimum: true`.

Until the the OpenAPI YAML is updated, the changes are available on the draft branch:

- https://forge.3gpp.org/rep/all/5G_APIs/-/tree/Rel17-draft-TSG102

**Comment by devbbc** [comment]
Sample outputs after uplifting the AF at reference point M1.

# Policy template resource: Optional applicationSessionContext

Since applicationSessionContext is now an optional property, the AF can create Policy Templates with and without this property.
For testing, [PolicyTemplate-1.json](https://github.com/5G-MAG/rt-5gms-application-function/blob/development/examples/PolicyTemplate-1.json) was used.

## With applicationSessionContext
```
$ curl -H 'User-Agent: AF' -H 'Content-Type: application/json' -d @PolicyTemplate-1.json -v http://127.0.0.23:7777/3gpp-m1/v2/provisioning-sessions/3306168c-aa43-41ee-9e7e-315888bca701/policy-templates
 
> POST /3gpp-m1/v2/provisioning-sessions/3306168c-aa43-41ee-9e7e-315888bca701/policy-templates HTTP/1.1
> Host: 127.0.0.23:7777
> Accept: */*
> User-Agent: AF
> Content-Type: application/json
> Content-Length: 520
>
 
< HTTP/1.1 201 Created
< Connection: Keep-Alive
< Content-Length: 0
< Location: /3gpp-m1/v2/provisioning-sessions/3306168c-aa43-41ee-9e7e-315888bca701/policy-templates/5bc7ea1e-aa43-41ee-9e7e-315888bca701
< Server: 5GMSAF-af.localdomain/17 (info.title=M1_PolicyTemplatesProvisioning; info.version=2.2.1) rt-5gms-application-function/1.4.0
< Date: Wed, 03 Jan 2024 14:21:17 GMT
```
_Retrieve the created policy template_

```
$ curl -H 'User-Agent: AF' -v http://127.0.0.23:7777/3gpp-m1/v2/provisioning-sessions/3306168c-aa43-41ee-9e7e-315888bca701/policy-templates/5bc7ea1e-aa43-41ee-9e7e-315888bca701
 
> GET /3gpp-m1/v2/provisioning-sessions/3306168c-aa43-41ee-9e7e-315888bca701/policy-templates/5bc7ea1e-aa43-41ee-9e7e-315888bca701 HTTP/1.1
> Host: 127.0.0.23:7777
> Accept: */*
> User-Agent: AF
>
 
< HTTP/1.1 200 OK
< Connection: Keep-Alive
< Content-Length: 606
< Cache-Control: max-age=60
< Last-Modified: Wed, 03 Jan 2024 14:21:17 GMT
< ETag: 569e1ff25f4c9956eeec0f1db3e42d4a67ec87ece58914a1a14b29dadc50e574
< Content-Type: application/json
< Server: 5GMSAF-af.localdomain/17 (info.title=M1_PolicyTemplatesProvisioning; info.version=2.2.1) rt-5gms-application-function/1.4.0
< Date: Wed, 03 Jan 2024 14:21:50 GMT
<
{
    "applicationSessionContext":    {
        "dnn":  "internet",
        "sliceInfo":    {
            "sd":   "000001",
            "sst":  1
        }
    },
    "chargingSpecification":    {
        "gpsi": ["msimsi-1234567890"],
        "sponId":   "sponsor-id-1",
        "sponStatus":   "SPONSOR_ENABLED"
    },
    "externalReference":    "policy-external-ref-1",
    "policyTemplateId": "5bc7ea1e-aa43-41ee-9e7e-315888bca701",
    "qoSSpecification": {
        "defPacketLossRateDl":  0,
        "defPacketLossRateUl":  0,
        "maxAuthBtrDl": "20 Mbps",
        "maxAuthBtrUl": "200 Kbps",
        "maxBtrDl": "10 Mbps",
        "maxBtrUl": "100 Kbps",
        "qosReference": "qos-1"
    },
    "state":    "READY",
    "stateReason":  {
    }
}
```
## Policy template without applicationSessionContext

For this test, _applicationSessionContext_ was removed from `PolicyTemplate-1.json`

```
$  curl -H 'User-Agent: AF' H 'Content-Type: application/json' -d @PolicyTemplate-1-without-appsessioncontext.json -v http://127.0.0.23:7777/3gpp-m1/v2/provisioning-sessions/3306168c-aa43-41ee-9e7e-315888bca701/policy-templates
 
> POST /3gpp-m1/v2/provisioning-sessions/3306168c-aa43-41ee-9e7e-315888bca701/policy-templates HTTP/1.1
> Host: 127.0.0.23:7777
> Accept: */*
> User-Agent: AF
> Content-Type: application/json
> Content-Length: 404
>
 
< HTTP/1.1 201 Created
< Connection: Keep-Alive
< Content-Length: 0
< Location: /3gpp-m1/v2/provisioning-sessions/3306168c-aa43-41ee-9e7e-315888bca701/policy-templates/79f1a160-aa43-41ee-9e7e-315888bca701
< Server: 5GMSAF-af.localdomain/17 (info.title=M1_PolicyTemplatesProvisioning; info.version=2.2.1) rt-5gms-application-function/1.4.0
< Date: Wed, 03 Jan 2024 14:22:08 GMT
```
_Retrieve the created policy template_

```
$ curl -H 'User-Agent: AF' -v http://127.0.0.23:7777/3gpp-m1/v2/provisioning-sessions/3306168c-aa43-41ee-9e7e-315888bca701/policy-templates/79f1a160-aa43-41ee-9e7e-315888bca701
 
> GET /3gpp-m1/v2/provisioning-sessions/3306168c-aa43-41ee-9e7e-315888bca701/policy-templates/79f1a160-aa43-41ee-9e7e-315888bca701 HTTP/1.1
> Host: 127.0.0.23:7777
> Accept: */*
> User-Agent: AF
>
 
< HTTP/1.1 200 OK
< Connection: Keep-Alive
< Content-Length: 497
< Server: 5GMSAF-af.localdomain/17 (info.title=M1_PolicyTemplatesProvisioning; info.version=2.2.1) rt-5gms-application-function/1.4.0
< Cache-Control: max-age=60
< Last-Modified: Wed, 03 Jan 2024 14:22:08 GMT
< ETag: c4aa320781076f512f88acd5bd5c62178f412199d529036e3c28a5339d0c526b
< Content-Type: application/json
< Date: Wed, 03 Jan 2024 14:22:24 GMT
<
{
    "chargingSpecification":    {
        "gpsi": ["msimsi-1234567890"],
        "sponId":   "sponsor-id-1",
        "sponStatus":   "SPONSOR_ENABLED"
    },
    "externalReference":    "policy-external-ref-2",
    "policyTemplateId": "79f1a160-aa43-41ee-9e7e-315888bca701",
    "qoSSpecification": {
        "defPacketLossRateDl":  0,
        "defPacketLossRateUl":  0,
        "maxAuthBtrDl": "20 Mbps",
        "maxAuthBtrUl": "200 Kbps",
        "maxBtrDl": "10 Mbps",
        "maxBtrUl": "100 Kbps",
        "qosReference": "qos-1"
    },
    "state":    "READY",
    "stateReason":  {
    }
}
```

# Consumption Reporting Provisioning API

Example valid Consumption Reporting Configuration:

```
{
        "reportingInterval":    10,
        "serverAddresses":      ["http://127.0.0.22:7778/3gpp-m5/v2/"],
        "locationReporting":    false,
        "accessReporting":      false,
        "samplePercentage":     100
}
```
Provisioning the valid Consumption Reporting Configuration:
```
$ curl -H 'User-Agent: AF' -H 'Content-Type: application/json' -d @consumption-report.json -v http://127.0.0.23:7777/3gpp-m1/v2/provisioning-sessions/e4b5ca04-ab19-41ee-9ad9-5d991b5dae62/consumption-reporting-configuration

> POST /3gpp-m1/v2/provisioning-sessions/e4b5ca04-ab19-41ee-9ad9-5d991b5dae62/consumption-reporting-configuration HTTP/1.1
> Host: 127.0.0.23:7777
> Accept: */*
> User-Agent: AF
> Content-Type: application/json
> Content-Length: 164
>

< HTTP/1.1 204 No Content
< Connection: Keep-Alive
< Server: 5GMSAF-af.localdomain/17 (info.title=M1_ConsumptionReportingProvisioning; info.version=2.1.0) rt-5gms-application-function/1.4.0
< Date: Thu, 04 Jan 2024 15:57:26 GMT
<
```
Trying to provision an invalid Consumption Reporting Configuration with `reportingInterval: 0`

```
$ curl -H 'User-Agent: AF' -H 'Content-Type: application/json' -d @consumption-report.json -v http://127.0.0.23:7777/3gpp-m1/v2/provisioning-sessions/e4b5ca04-ab19-41ee-9ad9-5d991b5dae62/consumption-reporting-configuration
 
> POST /3gpp-m1/v2/provisioning-sessions/e4b5ca04-ab19-41ee-9ad9-5d991b5dae62/consumption-reporting-configuration HTTP/1.1
> Host: 127.0.0.23:7777
> Accept: */*
> User-Agent: AF
> Content-Type: application/json
> Content-Length: 163
>
* upload completely sent off: 163 out of 163 bytes
* Mark bundle as not supporting multiuse
< HTTP/1.1 400 Bad Request
< Connection: Keep-Alive
< Content-Length: 375
< Content-Type: application/problem+json
< Server: 5GMSAF-af.localdomain/17 (info.title=M1_ConsumptionReportingProvisioning; info.version=2.1.0) rt-5gms-application-function/1.4.0
< Date: Thu, 04 Jan 2024 15:57:46 GMT
<
{
    "type": "/3gpp-m1/v2",
    "title":    "Bad request.",
    "status":   400,
    "detail":   "Bad ConsumptionReportingConfiguration for provisioning session [e4b5ca04-ab19-41ee-9ad9-5d991b5dae62]: Field \"reportingInterval\" value is less than or equal to 0 [range (0..inf]]",
    "instance": "/provisioning-sessions/e4b5ca04-ab19-41ee-9ad9-5d991b5dae62/consumption-reporting-configuration"
}
```
# Returning error response code 400 Bad request 

Creating a content hosting configuration with no distribution_configurations returns 400 Bad Request.
```
$ curl -H 'User-Agent: AF' -H 'Content-Type: application/json' --data @ContentHostingConfiguration_AMP-Samples_pull-ingest.json -v http://127.0.0.23:7777/3gpp-m1/v2/provisioning-sessions/e4b5ca04-ab19-41ee-9ad9-5d991b5dae62/content-hosting-configuration
 
> POST /3gpp-m1/v2/provisioning-sessions/e4b5ca04-ab19-41ee-9ad9-5d991b5dae62/content-hosting-configuration HTTP/1.1
> Host: 127.0.0.23:7777
> Accept: */*
> User-Agent: AF
> Content-Type: application/json
> Content-Length: 260
>
 
< HTTP/1.1 400 Bad Request
< Connection: Keep-Alive
< Content-Length: 342
< Server: 5GMSAF-af.localdomain/17 (info.title=M1_ContentHostingProvisioning; info.version=2.2.3) rt-5gms-application-function/1.4.0
< Content-Type: application/problem+json
< Date: Thu, 04 Jan 2024 16:01:40 GMT
<
{
"type": "/3gpp-m1/v2",
"title": "Bad request.",
"status": 400,
"detail": "Bad ContentHostingConfiguration for provisioning session [e4b5ca04-ab19-41ee-9ad9-5d991b5dae62]: Required field \"distribution_configurations\" not found",
"instance": "/provisioning-sessions/e4b5ca04-ab19-41ee-9ad9-5d991b5dae62/content-hosting-configuration"
 
}
```

**Comment by davidjwbbc** [comment]
Closed by PR #118
